srcs/juloo.keyboard2/KeyboardData.java

```diff
@@ -40,7 +42,31 @@ public KeyboardData addExtraKeys(Iterator<KeyValue> k)
     addExtraKeys_to_row(rows, k, 1, 3);
     addExtraKeys_to_row(rows, k, 2, 2);
     addExtraKeys_to_row(rows, k, 2, 1);
-    return new KeyboardData(rows, keysWidth, extra_keys);
+    return new KeyboardData(rows, keysWidth, extra_keys, num_pad);
+  }
+
+  public KeyboardData addNumPad()
```

Comment:
"Blind" keys are a nice trick but I think they are not the best way. They show up all over the code and that's some subtle bugs in the future.

I think you can compute the needed "blind" space from here. The first key of each row of the numpad would need `this.keysWidth - row.keysWidth` of extra "shift".